eth-repository/src/main/kotlin/net/consensys/cava/eth/repository/BlockchainRepository.kt

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package net.consensys.cava.eth.repository
+
+import net.consensys.cava.bytes.Bytes
+import net.consensys.cava.bytes.Bytes32
+import net.consensys.cava.eth.Block
+import net.consensys.cava.eth.BlockHeader
+import net.consensys.cava.eth.Hash
+import net.consensys.cava.kv.KeyValueStore
+import java.nio.charset.StandardCharsets
+
+/**
+ * Repository housing blockchain information.
+ *
+ * This repository allows storing blocks, block headers and metadata about the blockchain, such as forks and head
+ * information.
+ */
+class BlockchainRepository
+/**
+ * Default constructor.
+ *
+ * @param chainMetadata the key-value store to store chain metadata
+ * @param blockStore the key-value store to store blocks
+ * @param blockHeaderStore the key-value store to store block headers
+ * @param blockchainIndex the blockchain index to index values
+ */
+  (
+    private val chainMetadata: KeyValueStore,
+    private val blockStore: KeyValueStore,
+    private val blockHeaderStore: KeyValueStore,
+    private val blockchainIndex: BlockchainIndex
+  ) {
+
+  companion object {
+
+    /**
+     * Initializes a blockchain repository with metadata, placing it in key-value stores.
+     *
+     * @return a new blockchain repository made from the metadata passed in parameter.
+     */
+    suspend fun init(
+      blockStore: KeyValueStore,
+      blockHeaderStore: KeyValueStore,
+      chainMetadata: KeyValueStore,
+      blockchainIndex: BlockchainIndex,
+      genesisBlock: Block
+    ): BlockchainRepository {
+      val repo = BlockchainRepository(chainMetadata, blockStore, blockHeaderStore, blockchainIndex)
+      repo.setGenesisBlock(genesisBlock)
+      repo.storeBlock(genesisBlock)
+      return repo
+    }
+  }
+
+  /**
+   * Stores a block into the repository.
+   *
+   * @param block the block to store
+   * @return a handle to the storage operation completion
+   */
+  suspend fun storeBlock(block: Block) {
+    blockStore.put(block.header().hash().toBytes(), block.toBytes())
+    blockHeaderStore.put(block.header().hash().toBytes(), block.header().toBytes())
```

Comment:
You _might_ want to improve this to run each `put` operation in parallel

Reply:
yeah, _maybe_. Thinking about it some more, since this is not really transactional (I'm not sure it's a big deal if we have orphaned values)